ql/src/java/org/apache/hadoop/hive/ql/udf/generic/AbstractGenericUDFArrayBase.java

```diff
@@ -0,0 +1,160 @@
+        }
+
+        arrayOI = (ListObjectInspector) arguments[ARRAY_IDX];
+        argumentOIs = arguments;
+
+        //return initialize(arguments);
+        return initListOI(arguments);
+    }
+
+    @Override
+    public String getDisplayString(String[] children) {
+        assert (children.length == MIN_ARG_COUNT);
+        return FUNC_NAME.toString().toLowerCase() + "(" + children[ARRAY_IDX] + ")";
+    }
+
+    List<Object> convertArray(List objects) {
+        List<Object> ret = new ArrayList<>();
+        for (Object o : objects) {
+            ret.add(converter.convert(o));
+        }
+        return ret;
+    }
+
+    void checkArgCategory(ObjectInspector[] arguments, int idx, Enum category,
+                          FUNC_NAMES function_name, String typeName) throws UDFArgumentTypeException {
+
+        if (!arguments[idx].getCategory().equals(category)) {
+            throw new UDFArgumentTypeException(idx,
+                    "\"" + typeName + "\" "
+                            + "expected at function " + function_name + ", but "
+                            + "\"" + arguments[idx].getTypeName() + "\" "
+                            + "is found");
+        }
+    }
+
+    void checkArgIntPrimitiveCategory(PrimitiveObjectInspector objectInspector,
+                                      FUNC_NAMES function_name, int idx) throws UDFArgumentTypeException {
+
+        switch (objectInspector.getPrimitiveCategory()) {
+            case SHORT:
+            case INT:
+            case LONG:
+                break;
+            default:
+                throw new UDFArgumentTypeException(0, "Argument " + idx
+                        + " of function " + function_name + " must be \""
+                        + serdeConstants.SMALLINT_TYPE_NAME + "\""
+                        + " or \"" + serdeConstants.INT_TYPE_NAME + "\""
+                        + " or \"" + serdeConstants.BIGINT_TYPE_NAME + "\", but \""
+                        + objectInspector.getTypeName() + "\" was found.");
+        }
+    }
+
+    boolean isListEmpty(Object array, ListObjectInspector listObjectInspector) {
+
+        int arrayLength = listObjectInspector.getListLength(array);
+
+        // Check if array is null or empty or value is null
+        return array == null || arrayLength <= 0;
+    }
+
+    ObjectInspector initListOI(ObjectInspector[] arguments) {
+
+        GenericUDFUtils.ReturnObjectInspectorResolver returnOIResolver =
+                new GenericUDFUtils.ReturnObjectInspectorResolver(true);
+
+        ObjectInspector elementObjectInspector =
+                ((ListObjectInspector) (arguments[0])).getListElementObjectInspector();
+
+        ObjectInspector returnOI = returnOIResolver.get(elementObjectInspector);
+        converter = ObjectInspectorConverters.getConverter(elementObjectInspector, returnOI);
+        if(FUNC_NAME == FUNC_NAMES.ARRAY_MAX || FUNC_NAME == FUNC_NAMES.ARRAY_MIN){
```

Comment:
Not sure how to handle this one yet, but again, if we can move these down to the derived class to avoid function names in the base class, I think that would be good.  Will think of a way later.